File: pkg/fixed128legacy/fixed128.go
package fixed128legacy

import (
	"fmt"
	"math/big"
)

var (
	ErrorDivisionByZero = fmt.Errorf("division by zero")
)

type Fixed128 struct {
	value big.Int
}

func New(x, y int64) (Fixed128, error) {
	var f128 Fixed128

	if y == 0 {
		return f128, ErrorDivisionByZero
	}

	absX, negX := normalize(x)
	absY, negY := normalize(y)
	neg := negX != negY

	xBig := big.NewInt(0).SetUint64(absX)
	yBig := big.NewInt(0).SetUint64(absY)

	f128.value.Lsh(xBig, 64)
	f128.value.Div(&f128.value, yBig)

	if neg {
		f128.value.Neg(&f128.value)
	}

	return f128, nil
}

func MustNew(x, y int64) Fixed128 {
	f128, err := New(x, y)
	if err != nil {
		panic(err)
	}
	return f128
}

func normalize(v int64) (uint64, bool) {
	mask := uint64(v >> 63)
	neg := mask != 0
	abs := (uint64(v) ^ mask) - mask
	return abs, neg
}

func (f128 Fixed128) Copy() Fixed128 {
	return Fixed128{
		value: *big.NewInt(0).Set(&f128.value),
	}
}

func (f128 Fixed128) Value() big.Int {
	return f128.Copy().value
}

func (f128 Fixed128) Sign() int {
	return f128.value.Sign()
}

func (f128 Fixed128) Cmp(other Fixed128) int {
	return f128.value.Cmp(&other.value)
}

func (f128 Fixed128) IsNeg() bool {
	return f128.Sign() < 0
}

func (f128 Fixed128) IsZero() bool {
	return f128.value.Sign() == 0
}

// func (f128 Fixed128) HiLo() (uint64, uint64) {
// 	_, hi, lo := disassemble(f128)
// 	return hi, lo
// }

// func (f128 Fixed128) Bytes() []byte {
// 	return f128.bytes()
// }

func (f128 Fixed128) Add(b Fixed128) Fixed128 {
	var result Fixed128
	result.value.Add(&f128.value, &b.value)
	return result
}

func (f128 Fixed128) Sub(b Fixed128) Fixed128 {
	var result Fixed128
	result.value.Sub(&f128.value, &b.value)
	return result
}

func (f128 Fixed128) Mul(b Fixed128) Fixed128 {
	var result Fixed128
	result.value.Mul(&f128.value, &b.value)
	return result
}

func (f128 Fixed128) Quo(b Fixed128) (Fixed128, error) {
	var result Fixed128
	result.value.Quo(&f128.value, &b.value)
	return result, nil
}

func (f128 Fixed128) Lsh(bits uint) Fixed128 {
	var result Fixed128
	result.value.Lsh(&f128.value, bits)
	return result
}

// func (f128 Fixed128) MulInt64(y int64) (int64, error) {
// 	return mulInt64(f128, y)
// }

File: Sources/Fixed128/Fixed128.swift
import Foundation

enum Fixed128Error: Error {
    case divisionByZero
}

public struct Fixed128: Equatable {
    public let neg: Bool
    public let hi: UInt64
    public let lo: UInt64

    public init(neg: Bool, hi: UInt64, lo: UInt64) {
        self.neg = neg
        self.hi = hi
        self.lo = lo
    }
}

public func NewFixed128(x: Int64, y: Int64) throws -> Fixed128 {
    return try toFixed128(x: x, y: y)
}

func toFixed128(x: Int64, y: Int64) throws -> Fixed128 {
    if y == 0 {
        throw Fixed128Error.divisionByZero
    }

    let (negX, absX) = normalize(value: x)
    let (negY, absY) = normalize(value: y)
    let neg = negX != negY

    let (hi, lo) = getComponents(x: absX, y: absY)

    return Fixed128(neg: neg, hi: hi, lo: lo)
}

func normalize(value: Int64) -> (neg: Bool, abs: UInt64) {
    let mask: UInt64 = UInt64(value >> 63)
    let neg: Bool = mask != 0
    let abs: UInt64 = (UInt64(value) ^ mask) - mask
    return (neg, abs)
}

func getComponents(x: UInt64, y: UInt64) -> (UInt64, UInt64) {
    var hi: UInt64 = 0
    var lo: UInt64 = 0

    hi = x / y
    var part = x % y

    let shift = y.leadingZeroBitCount
    var div = y << shift
    part <<= shift

    var i = 0
    while i < 64 && div > 1 && part > 0 {
        div >>= 1
        let bit = 1 & ~UInt64(bitPattern: Int64(bitPattern: part - y) >> 63)
        part -= bit * div
        lo <<= 1
        lo |= bit
        i += 1
    }

    lo <<= (64 - i)
    return (hi, lo)
}
